File: src/python/cytoprofiling/filtering.py
import numpy as np
import pandas as pd
from cytoprofiling.utility import get_wells, get_barcoding_batches, get_default_normalization_targets

def down_sample_cells(df : pd.DataFrame, target_count : int) -> pd.DataFrame:
    """Down sample to fixed number of cells per well
    
    Args:
        df : Input dataframe with cells
        target_count : Target number of cells per well

    Returns:
        Down sampled data frame
    """

    keep = np.zeros(df.shape[0])
    for well in get_wells(df):
        num_cells = np.sum(df["Well"] == well)
        cell_target = min(num_cells, target_count)
        cell_keep = ([1,] * cell_target) + ([0,] * (num_cells - cell_target))
        np.random.shuffle(cell_keep)
        keep[df["Well"] == well] = cell_keep
    
    result = df.loc[keep == 1,:]
    result.reset_index(drop=True, inplace=True)
    return result


def filter_cells_by_area_percentile(df : pd.DataFrame, well_names : list[str] = None, min_percentile : float = 3, max_percentile : float = 97) -> pd.DataFrame:
    """Filter cells to remove outliers by area

    Args:
      df : Input dataframe with cells
      well_names : List of wells to include in filtering
      min_percentile : Percentile to determine minimum value for thresholding
      max_percentile : Percentile to determine maximum value for thresholding

    Returns:
      Filtered data frame
    """
    if well_names is None:
        well_names = get_wells(df)
    filter = np.zeros(df.shape[0], dtype=np.int8)
    for well_name in well_names:
        well_areas = df[df["Well"] == well_name]["Area"]
        min_threshold = np.nanpercentile(well_areas, min_percentile) - 0.001
        max_threshold = np.nanpercentile(well_areas, max_percentile) + 0.001

        filter[(df["Well"] == well_name)&(well_areas >= min_threshold)&(well_areas <= max_threshold)] = 1
    
    result = df.loc[filter > 0,:]
    result.reset_index(drop=True, inplace=True)
    return result


def filter_cells_by_assigned_counts_percentile(df : pd.DataFrame, batch_names : list[str] = None, well_names : list[str] = None, min_percentile : float = 3, max_percentile : float = 97) -> pd.DataFrame:
    """Filter cells to remove outliers by assigned counts in a batch

    Cells must fall within percentile range for all batches to be kept

    Args:
      df : Input dataframe with cells
      batch_name : List of batches to include in filtering
      well_names : List of wells to include in filtering
      min_percentile : Percentile to determine minimum value for thresholding
      max_percentile : Percentile to determine maximum value for thresholding

    Returns:
      Filtered data frame
    """
    if well_names is None:
        well_names = get_wells(df)
    if batch_names is None:
        batch_names = get_barcoding_batches(df)

    filters = []
    for batch_name in batch_names:
        batch_columns = get_default_normalization_targets(df, batch_name)
        assigned_counts = np.nansum(df[batch_columns], axis = 1)
        filter = np.zeros(df.shape[0], dtype=np.uint8)

        for well_name in well_names:
            well_counts = assigned_counts[(df["Well"] == well_name)&(assigned_counts > 0)]
            min_threshold = np.nanpercentile(well_counts, min_percentile) - 0.001
            max_threshold = np.nanpercentile(well_counts, max_percentile) + 0.001
            filter[(df["Well"] == well_name)&(assigned_counts > min_threshold)&(assigned_counts<max_threshold)] = 1
        filters.append(filter)

    # compute the intersection of all filters
    filter = np.ones(df.shape[0], dtype=np.uint8)
    for f in filters:
        filter = filter & f

    result = df.loc[filter > 0,:]
    result.reset_index(drop=True, inplace=True)
    return result

def filter_cells_by_assigned_rate(df : pd.DataFrame, batch_names : list[str] = None, well_names : list[str] = None, assigned_rate : float = 0.5) -> pd.DataFrame:
    """Filter cells on assigned rate

    Cells must exceed the assignment rate threshold in all batches
    to be kept

    Args:
      df : Input dataframe with cells
      batch_name : List of batches to use for filtering
      well_names : List of wells to use in filtering
      assigned_rate : Assigned rate to use for filtering cells

    Returns:
      Filtered data frame
    """
    if well_names is None:
        well_names = get_wells(df)
    if batch_names is None:
        batch_names = get_barcoding_batches(df)

    filters = []
    for batch_name in batch_names:
        batch_columns = get_default_normalization_targets(df, batch_name)
        assigned_counts = np.nansum(df[batch_columns], axis = 1)
        unassigned_counts = df[f"Unassigned.{batch_name}"]
        filter = np.zeros(df.shape[0], dtype=np.uint8)

        for well_name in well_names:
            filter[(df["Well"] == well_name)&(assigned_counts > 0)&(assigned_counts / ((unassigned_counts + assigned_counts)) > assigned_rate)] = 1
        filters.append(filter)

    # compute the intersection of all filters
    filter = np.ones(df.shape[0], dtype=np.uint8)
    for f in filters:
        filter = filter & f

    result = df.loc[filter > 0,:]

    result.reset_index(drop=True, inplace=True)
    return result

def filter_cells(df : pd.DataFrame, batch_names : list[str] = None, well_names : list[str] = None, stats = None) -> pd.DataFrame:
    """Perform default filtering of cells (filter by area, assigned counts, and assigned rate)

    Args:
      df : Input dataframe with cells
      batch_names : List of batches to use for normalization
      well_names : List of wells to include in normalization
      stats : Dictionary to store statistics about filtering under the key 'filter_cells'
      
    Returns:
      Filtered data frame
    """
    if well_names is None:
        well_names = get_wells(df)
    if batch_names is None:
        batch_names = get_barcoding_batches(df)
    if stats is None:
        stats = {}
    
    result = filter_cells_by_area_percentile(df, well_names)
    for well in well_names:
        stats.setdefault(well, []).append(np.sum(df["Well"] == well) - np.sum(result["Well"] == well))
    df = result
    result = filter_cells_by_assigned_counts_percentile(df, batch_names, well_names)
    for well in well_names:
        stats.setdefault(well, []).append(np.sum(df["Well"] == well) - np.sum(result["Well"] == well))
    df = result
    result = filter_cells_by_assigned_rate(df, batch_names, well_names, 0.5)
    for well in well_names:
        stats.setdefault(well, []).append(np.sum(df["Well"] == well) - np.sum(result["Well"] == well))
    for well in well_names:
        stats.setdefault(well, []).append(np.sum(result["Well"] == well))

    stats_df = pd.DataFrame.from_dict(stats)
    stats_df.index = ["Area", "AssignedCounts", "AssignedRate", "Passing"]
    stats["filter_cells"] = stats_df
    return result

File: src/R/cytoprofiling/R/filtering.R
#' Filter cells to remove outliers by area
#'
#' @param df Cell table
#' @param well_names List of well names
#' @param min_threshold Lower percentile threshold
#' @param max_threshold Upper percentile threshold
#' @return Filtered cell table
#' @export
filter_cells_by_area_percentile <- function(df, well_names = c(), min_threshold = 3, max_threshold = 97) {
  if (length(well_names) == 0) {
    well_names <- unique(df$Well)
  }
  filter_status <- rep(FALSE, dim(df)[[1]])
  for (well in well_names) {
    thresholds <- quantile(df[df$Well == well, ]$Area, c(min_threshold / 100.0, max_threshold / 100.0), na.rm = TRUE)
    filter_status <- filter_status | (df$Well == well) & (df$Area > thresholds[[1]]) & (df$Area < thresholds[[2]])
  }
  return(df[filter_status, ])
}

#' Filter cells to remove outliers by assigned counts
#'
#' Cells must fall within percentile range for all batches to be kept
#'
#' @param df Cell table
#' @param batch_names List of batch names
#' @param well_names List of well names
#' @param min_threshold Lower percentile threshold
#' @param max_threshold Upper percentile threshold
#' @return Filtered cell table
#' @export
filter_cells_by_assigned_counts_percentile <- function(df, batch_names = c(), well_names = c(), min_threshold = 3, max_threshold = 97) {
  if (length(batch_names) == 0) {
    batch_names <- get_barcoding_batches(df)
  }
  if (length(well_names) == 0) {
    well_names <- unique(df$Well)
  }

  final_filter_status <- rep(TRUE, dim(df)[[1]])

  for (batch in batch_names) {
    normalization_targets <- get_default_normalization_targets(df, batch)
    norm_values <- rowSums(df[, normalization_targets], na.rm = TRUE)

    current_filter_status <- rep(FALSE, dim(df)[[1]])
    for (well in well_names) {
      thresholds <- quantile(norm_values[(df$Well == well) & (norm_values > 0)], c(min_threshold / 100.0, max_threshold / 100.0), na.rm = TRUE)
      current_filter_status <- current_filter_status | (df$Well == well) & (norm_values > thresholds[[1]]) & (norm_values < thresholds[[2]])
    }
    final_filter_status <- final_filter_status & current_filter_status
  }
  return(df[final_filter_status, ])
}



#' Filter cells on assigned rate
#'
#' Cells must exceed the assignment rate threshold in all batches
#' to be kept
#'
#' @param df Cell table
#' @param batch_names List of batch names
#' @param well_names List of well names
#' @param assigned_rate_threshold Assigned rate threshold
#' @return Filtered cell table
#' @export
filter_cells_by_assigned_rate <- function(df, batch_names = c(), well_names = c(), assigned_rate_threshold = 0.5) {
  if (length(batch_names) == 0) {
    batch_names <- get_barcoding_batches(df)
  }
  if (length(well_names) == 0) {
    well_names <- unique(df$Well)
  }

  final_filter_status <- rep(TRUE, dim(df)[[1]])
  for (batch_name in batch_names) {
    batch_columns <- get_default_normalization_targets(df, batch_name)
    assigned_counts <- rowSums(df[, batch_columns], na.rm = TRUE)
    unassigned_counts <- df[[paste("Unassigned.", batch_name, sep = "")]]
    current_filter_status <- rep(FALSE, dim(df)[[1]])
    for (well_name in well_names) {
      current_filter_status <- current_filter_status | ((df$Well == well_name) & (assigned_counts > 0) & (assigned_counts / (unassigned_counts + assigned_counts) > assigned_rate_threshold))
    }
    final_filter_status <- final_filter_status & current_filter_status
  }
  return(df[final_filter_status, ])
}

#' Perform default filtering of cells (filter by area, assigned counts, and assigned rate)
#'
#' @param df Cell table
#' @return Filtered cell table
#' @export
filter_cells <- function(df, batch_names = c(), well_names = c()) {
  return(filter_cells_by_assigned_rate(
    filter_cells_by_area_percentile(
      filter_cells_by_assigned_counts_percentile(
        df,
        batch_names = batch_names,
        well_names = well_names
      ),
      well_names = well_names
    ),
    batch_names = batch_names,
    well_names = well_names
  ))
}
